### PR TITLE
PHPUnit & browser tests for testing locally and on Github Actions

### PR DESCRIPTION
**Pull request by addshore**

### CODE REVIEW

**Comment by addshore** [comment]
I should probably also write some docs and add more examples before this is merged...

**Review by oshotton — COMMENTED** [review]
Looks good. 
Haven't been able to test this locally though.
Running `docker-compose -f docker-compose.yml -f docker-compose-selenium.yml up` followed by `docker run -it --rm -v %CD%:/app composer phpunit:browser` (in a second terminal) errors like so:
```
$ docker run -it --rm -v %CD%:/app composer phpunit:browser
> phpunit tests/browser
PHPUnit 9.5.0 by Sebastian Bergmann and contributors.

E                                                                   1 / 1 (100%)

Time: 00:00.354, Memory: 4.00 MB

There was 1 error:

1) SWA\Test\Browser\TitleTest::testTitle
Facebook\WebDriver\Exception\WebDriverCurlException: Curl error thrown for http POST to /session with params: {"desiredCapabilities":{"browserName":"chrome","platform":"ANY","chromeOptions":{"w3c":false,"binary":""}}}

Failed to connect to localhost port 4444: Connection refused

/app/vendor/facebook/webdriver/lib/Remote/HttpCommandExecutor.php:297
/app/vendor/facebook/webdriver/lib/Remote/RemoteWebDriver.php:144
/app/tests/browser/SWABrowserTestCase.php:32
/app/tests/browser/SWABrowserTestCase.php:39
/app/tests/browser/SWABrowserTestCase.php:55

ERRORS!
Tests: 1, Assertions: 0, Errors: 1.
Script phpunit tests/browser handling the phpunit:browser event returned with error code 2
```

File: docker-compose-selenium.yml [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+  #     - ./.docker/selenium/videos:/videos
+  #   depends_on:
+  #     - opera
+  #   environment:
+  #     - DISPLAY_CONTAINER_NAME=opera
+  #     - FILE_NAME=opera_video.mp4
+
+  selenium-hub:
+    image: selenium/hub:4.0.0-beta-1-prerelease-20201208
+    # container_name: selenium-hub
```
[the remark this answers is not in the document]
Comment (addshore) [reply docker-compose-selenium.yml:78]: So this doesn't appear to be needed.

File: tests/browser/SWABrowserTestCase.php [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+				$this->setWebDriverWithRetries($attempts);
+			}
+		}
+	}
+
+	public function setUp() : void
+    {
+		$this->setWebDriverWithRetries();
+        $this->setBaseUrl();
+	}
```
[the remark this answers is not in the document]
Comment (addshore) [reply tests/browser/SWABrowserTestCase.php:57]: Yeah, the phpcs fixer actually ended up causing this formatting, it technically meets all of the rules :P

File: docker-compose-selenium.yml (hunk 1 above)
[the remark this answers is not in the document]
Comment (outdooracorn) [reply docker-compose-selenium.yml:78]: Right! Let's leave it in. Could add a comment to say why it's left in but commented out?

[the remark this answers is not in the document]
Comment (oshotton) [reply docker-compose-selenium.yml:78]: Right! Let's leave it in. Could add a comment to say why it's left in but commented out?

File: tests/browser/SWABrowserTestCase.php (hunk 2 above)
[the remark this answers is not in the document]
Comment (addshore) [reply tests/browser/SWABrowserTestCase.php:57]: Looks like this was my IDE fighting against the code style :P

File: docker-compose-selenium.yml (hunk 1 above)
[the remark this answers is not in the document]
Comment (addshore) [reply docker-compose-selenium.yml:84]: DOne!

**Comment by addshore** [comment]
> Looks good.
> Haven't been able to test this locally though.

This should now work.

Docs also done, so this is probably ready to be merged as is IMO, and then we can move forward slowly adding tests etc

File: README.md [hunk 3]
```diff
@@ -27,7 +27,45 @@ add the code snippet from https://github.com/SWAuk/plg_swa_viewlevels
...
+### composer
+
+In order to install the needed packages for development you'll need [composer](https://getcomposer.org/download/).
+Composer is a dependency manager and command runner for PHP.
+
+Use it to install the PHP libraries needed for testing, development and production.
+
+As docker is already required, the easiest way to use `composer` to install the required libs is via the official [docker image](https://hub.docker.com/_/composer).
+
+You can also run **composer in docker**...
```
Comment (oshotton) [inline README.md:41]: This line now not needed as the line above explains it better?

File: README.md [hunk 4]
```diff
@@ -27,7 +27,45 @@ add the code snippet from https://github.com/SWAuk/plg_swa_viewlevels
...
+```sh
+docker run --rm -it -v %CD%:/app composer <command here>
+```
+
+### docker & docker-compose
+
+If you want to run composer in docker, or use the recommended development environment then you need these installed.
+
+* [Docker](https://docs.docker.com/install/)
+* [docker-compose](https://docs.docker.com/compose/install/)
```
Comment (oshotton) [inline README.md:60]: Move this section above the composer section? Then it talks about installing docker before using docker.

**Review by oshotton — APPROVED** [review]
Looks good! :)

File: README.md (hunk 4 above)
Comment (addshore) [reply README.md:60]: Will do

File: README.md (hunk 3 above)
Comment (addshore) [reply README.md:41]: Will do

**Pull request merged**